lib/java/ydoc-server/src/main/java/org/enso/ydoc/polyfill/web/Zlib.java

```diff
@@ -0,0 +1,135 @@
+package org.enso.ydoc.polyfill.web;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+import java.util.zip.DeflaterOutputStream;
+import java.util.zip.InflaterOutputStream;
+import org.enso.ydoc.Polyfill;
+import org.enso.ydoc.polyfill.Arguments;
+import org.graalvm.polyglot.Context;
+import org.graalvm.polyglot.Source;
+import org.graalvm.polyglot.Value;
+import org.graalvm.polyglot.io.ByteSequence;
+import org.graalvm.polyglot.proxy.ProxyExecutable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Implements the <a href="https://nodejs.org/api/zlib.html">Zlib</a> Node.js interface. */
+final class Zlib implements Polyfill, ProxyExecutable {
+
+  private static final Logger log = LoggerFactory.getLogger(Zlib.class);
+
+  private static final String BUFFER_FROM = "buffer-from";
+  private static final String BUFFER_TO_STRING = "buffer-to-string";
+  private static final String ENCODING_BASE64 = "base64";
+  private static final String ENCODING_BASE64_URL = "base64url";
+
+  private static final String ZLIB_DEFLATE_SYNC = "zlib-deflate-sync";
+  private static final String ZLIB_INFLATE_SYNC = "zlib-inflate-sync";
+
+  private static final String ZLIB_JS = "zlib.js";
+
+  @Override
+  public void initialize(Context ctx) {
+    final var jsSource = Source.newBuilder("js", getClass().getResource(ZLIB_JS)).buildLiteral();
+
+    ctx.eval(jsSource).execute(this);
+  }
+
+  @Override
+  public Object execute(Value... arguments) {
+    final var command = arguments[0].asString();
+
+    log.debug(Arguments.toString(arguments));
+
+    return switch (command) {
+      case BUFFER_FROM -> {
+        final var text = arguments[1].asString();
+        final var encoding = arguments[2].asString();
+
+        yield switch (encoding) {
+          case ENCODING_BASE64 -> {
+            final var buffer = StandardCharsets.UTF_8.encode(text);
+            yield Base64.getDecoder().decode(buffer);
+          }
+          case ENCODING_BASE64_URL -> {
+            final var buffer = StandardCharsets.UTF_8.encode(text);
+            yield Base64.getUrlDecoder().decode(buffer);
+          }
+          case null -> StandardCharsets.UTF_8.encode(text);
+          default -> {
+            Charset charset;
+            try {
+              charset = Charset.forName(encoding);
+            } catch (IllegalArgumentException ignored) {
```

Comment:
Sounds like we should log this at least.

Reply:
Made it compatible with the nodejs by returning an error.